tests/unittests/tests-nanocoap/tests-nanocoap.c

```diff
@@ -334,7 +334,56 @@ static void test_nanocoap__get_multi_query(void)
     /* skip initial '&' from coap_get_uri_query() */
     TEST_ASSERT_EQUAL_STRING((char *)qs, &query[1]);
 }
+/*
+ * Builds on get_multi_query test, to use coap_opt_add_uquery2().
+ */
+static void test_nanocoap__add_uquery2(void)
+{
+    uint8_t buf[_BUF_SIZE];
+    coap_pkt_t pkt;
+    uint16_t msgid = 0xABCD;
+    uint8_t token[2] = {0xDA, 0xEC};
+    char keys[] = "a;bcd;";
+    int key1_len = 1;
+    int key2_len = 3;
+    char vals[] = "do;re";
+    int val1_len = 2;
+    char qs1[] = "a=do";
+    size_t query1_opt_len = 6;    /* first opt header is 2 bytes long */
+    char qs2[] = "a=do&bcd";
+    size_t query2_opt_len = 4;
+    char qs3[] = "a=do&bcd&bcd";
+    size_t query3_opt_len = 4;
+
+    size_t len = coap_build_hdr((coap_hdr_t *)&buf[0], COAP_TYPE_NON,
+                                &token[0], 2, COAP_METHOD_GET, msgid);
+
+    coap_pkt_init(&pkt, &buf[0], sizeof(buf), len);
+
+    /* includes key and value */
+    char query[20] = {0};
+    len = coap_opt_add_uquery2(&pkt, keys, key1_len, vals, val1_len);
+    TEST_ASSERT_EQUAL_INT(query1_opt_len, len);
+    coap_get_uri_query(&pkt, (uint8_t *)&query[0]);
+    /* skip initial '&' from coap_get_uri_query() */
+    TEST_ASSERT_EQUAL_STRING((char *)qs1, &query[1]);
 
+    /* includes key only */
+    memset(query, 0, 20);
+    len = coap_opt_add_uquery2(&pkt, &keys[2], key2_len, NULL, 0);
+    TEST_ASSERT_EQUAL_INT(query2_opt_len, len);
+    coap_get_uri_query(&pkt, (uint8_t *)&query[0]);
+    /* skip initial '&' from coap_get_uri_query() */
+    TEST_ASSERT_EQUAL_STRING((char *)qs2, &query[1]);
+
+    /* includes key only; value not NULL but zero length */
```

Comment:
How about test for value NULL but length not zero? The current implementation will ignore value length.

Reply:
Added; runs only when asserts disabled.